### ISSUE TITLE
some questions about Response_cache

### ISSUE THREAD

**Issue by wlzzZ** [horovod/horovod#1305]
In the following code, if the message is in the local response cache but not in coordinator cache, you choose to process again in the next cycle. If the tensor is ready to send the rank=0, I think it's better to send it to the rank=0 as earlier as possible. when the tensor is not in the response cache, it is hard to avoid having to send the request. 

https://github.com/horovod/horovod/blob/02329840846c5563537265b7b058e24fbd301044/horovod/common/controller.cc#L128

**Comment by romerojosh** [comment]
The purpose of the `response_cache` is to avoid redundantly sending request information to the coordinator (i.e. rank 0) over the network for scheduling redundantly every training step. When a request corresponding to a cached response is encountered on a worker, it must know that all other workers also have that request ready during the same cycle (accomplished via a bitvector reduction that occurs as part of the sync method here https://github.com/horovod/horovod/blob/02329840846c5563537265b7b058e24fbd301044/horovod/common/response_cache.cc#L307). This bitvector reduction enables all workers to make a collective decision on whether a request is globally ready for processing or not without the need for a coordinator rank. As a result, workers must hold back requests that have been determined to not be ready on all workers in a given cycle. 

This is actually no different than what would happen if we used the coordinator. A worker sends request information to be recorded by the coordinator in the first cycle it sees the request regardless of the state of other workers; however, the coordinator will in turn hold back any additional processing until the cycle it has recorded a common request from all workers. The cached path just makes this postponement of processing more obvious.

**Comment by wlzzZ** [comment]
The cache path just makes this postponement of processing more obvious, which means I close the cache path and could get a higher efficiency?

**Comment by wlzzZ** [comment]
I still have a question about the cache update method. The response cache is adopted LRU. In each training step,  mostly a tensor's response would only broadcast once? If one receives a tensor's response, it may wait until the next training step and then send the ready tensor again. The LRU does not seem to improve efficiency.

**Comment by romerojosh** [comment]
The purpose of the response cache is to improve the efficiency of the control plane, the phase of the code that coordinates the worker collective communication, _not_ reduce the amount of tensor data sent for the actual collective communication (e.g. allreduce). The original uncached coordination strategy relies on sending many auxiliary MPI messages to the coordinator rank every step, which is inefficient. The cached strategy removes the need to send these extra messages every step just to coordinate workers. Reading through #902 may clear this up further.

**Comment by wlzzZ** [comment]
if I set the size of response_cache 0 or some values(larger than the number of tensors), the Img/sec nearly are same. But if I set the response_cache size between 0 and the number of tensors, it will have a poor performance. I test it in one machine with multi-GPUs. Does it mean that if I train the model in one machine with multi-GPUs, the efficiency can not benefit from the response cache?

**Comment by wlzzZ** [comment]
[In response_cache.cc, this warning ](https://github.com/horovod/horovod/blob/a639de51e9a38d5c1f99f458c045aeaebe70351e/horovod/common/response_cache.cc#L112l)

**Comment by wlzzZ** [comment]
Thanks for your patient comments. I still have a question that whether the horovod supports gradient compression.

**Comment by abditag2** [comment]
Yes it does. You can specify the compression type in the DistributedOptimizer class.

**Comment by abditag2** [comment]
You can also create your own compression scheme.

**Comment by wlzzZ** [comment]
I mean that the sparse expression of  gradients, you can see this [paper](https://openreview.net/pdf?id=SkhQHMW0W)

**Comment by abditag2** [comment]
We don't have this compression implemented. But we encourage contributions if you are willing to implement it we can gladly help you. The changes we have made in the past few months provides clean interfaces for implementing new compression schemes.

**Comment by huoliquankai7** [comment]
What is an extra mpi message? If the tensor of each training step is the same, does that mean that it does not need to be controlled by the coordinator (such as rank0)? Take the local response directly to perform allreduce?

**Comment by huoliquankai7** [comment]
@tgaddair ，Thank you for your prompt reply to help me solve the doubts.

**Comment by tgaddair** [comment]
You're welcome.  I'm going to close this for now.
